components/constellation/network_listener.rs

```diff
@@ -0,0 +1,138 @@
+//! response is received, it is forwarded to the appropriate script thread.
+
+use hyper::header::Location;
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+use msg::constellation_msg::PipelineId;
+use net::http_loader::{set_default_accept, set_default_accept_language};
+use net_traits::{CoreResourceMsg, FetchMetadata, FetchResponseMsg};
+use net_traits::{IpcSend, NetworkError, ResourceThreads};
+use net_traits::request::{Destination, Referrer, RequestInit, Type};
+use net_traits::response::ResponseInit;
+use std::sync::mpsc::Sender;
+
+pub struct NetworkListener {
+    res_init: Option<ResponseInit>,
+    req_init: RequestInit,
+    pipeline_id: PipelineId,
+    resource_threads: ResourceThreads,
+    sender: Sender<(PipelineId, FetchResponseMsg)>,
+    should_send: bool,
+}
+
+impl NetworkListener {
+    pub fn new(req_init: RequestInit,
+               pipeline_id: PipelineId,
+               resource_threads: ResourceThreads,
+               sender: Sender<(PipelineId, FetchResponseMsg)>) -> NetworkListener {
+        NetworkListener {
+            res_init: None,
+            req_init,
+            pipeline_id,
+            resource_threads,
+            sender,
+            should_send: false
+        }
+    }
+
+    pub fn initiate_fetch(&self) {
+        let (ipc_sender, ipc_receiver) = ipc::channel().expect("Failed to create IPC channel!");
+
+        let mut listener = NetworkListener {
+            res_init: self.res_init.clone(),
+            req_init: self.req_init.clone(),
+            resource_threads: self.resource_threads.clone(),
+            sender: self.sender.clone(),
+            pipeline_id: self.pipeline_id.clone(),
+            should_send: false,
+        };
+
+        ROUTER.add_route(ipc_receiver.to_opaque(), box move |message| {
+            let msg = message.to();
+            match msg {
+                Ok(FetchResponseMsg::ProcessRequestReferrer(referrer, referrer_policy)) => {
+                    listener.req_init.referrer_url = match referrer.clone() {
+                        Referrer::ReferrerUrl(servo_url) => Some(servo_url),
+                        _ => None,
+                    };
+                    listener.req_init.referrer_policy = referrer_policy;
+                }
+                Ok(FetchResponseMsg::ProcessResponse(res)) => listener.check_redirect(res),
+                Ok(msg_) => listener.send(msg_),
+                Err(e) => warn!("Error while receiving network listener message: {}", e),
+            };
+        });
+
+        let msg = match self.res_init {
+            Some(ref res_init_) => CoreResourceMsg::FetchRedirect(
+                                   self.req_init.clone(),
+                                   res_init_.clone(),
+                                   ipc_sender),
+            None => {
+                let mut req_init = self.req_init.clone();
+                set_default_accept(Type::None, Destination::Document, &mut req_init.headers);
```

Comment:
I didn't feel the need to create another function for this here, as this one and the one below seem to work perfectly, and it doesn't feel like a hack.